docs/src/main/asciidoc/qute.adoc

```diff
@@ -0,0 +1,271 @@
+.ItemResource.java
+[source,java]
+----
+package org.acme.quarkus.sample;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import io.quarkus.qute.Rendering;
+import io.quarkus.qute.Template;
+
+@Path("item")
+public class ItemResource {
+
+    @Inject
+    ItemService service;
+
+    @Inject
+    Template item; <1>
+    
+    @GET
+    @Path("{id}")
+    @Produces(MediaType.TEXT_HTML)
+    public Rendering get(@PathParam("id") Integer id) {
+        return item.render().data("item", service.findItem(id)); <2>
+    }
+
+    @TemplateExtension <3>
+    static BigDecimal discountedPrice(Item item) {
+        return item.price.multiply(new BigDecimal("0.9"));
+    }
+}
+----
+<1> Inject the template with path `META-INF/resources/templates/item.html`.
+<2> Make the `Item` object accessible in the template.
+<3> A static template extension method can be used to add "computed properties" to a data class. The class of the first parameter is used to match the base object and the method name is used to match the property name.
+
+== Rendering Periodic Reports
+
+Templating engine could be also very useful when rendering periodic reports.
+You'll need to add `quarkus-scheduler` and `quarkus-qute` extensions first.
+In your `pom.xml` file, add:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-qute</artifactId>
+</dependency>
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-scheduler</artifactId>
+</dependency>
+----
+
+Let's suppose the have a `SampleService` bean whose `get()` method returns a list of samples.
+
+.Sample.java
+[source,java]
+----
+public class Sample {
+    public boolean valid;
+    public String name;
+    public String data;
+}
+----
+
+The template is simple:
+
+.report.html
+[source,html]
+----
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>Report {now}</title>
+</head>
+<body>
+    <h1>Report {now}</h1>
+    {#for sample in samples} <1>
+      <h2>{sample.name ?: 'Unknown'}</h2> <2>
+      <p>
+      {#if sample.valid}
+        {sample.data}
+      {#else}
+        <strong>Invalid sample found</strong>.
+      {/if}
+      </p>
+    {/for}
+</body>
+</html>
+----
+<1> The loop section makes it possible to iterate over iterables, maps and streams.
+<2> This value expression is using elvis notation - if name is null the default value is used.
+
+[source,java]
+.ReportGenerator.java
+----
+package org.acme.quarkus.sample;
+
+import javax.inject.Inject;
+
+import io.quarkus.qute.Template;
+import io.quarkus.qute.api.ResourcePath;
+import io.quarkus.scheduler.Scheduled;
+
+public class ReportGenerator {
+
+    @Inject
+    SampleService service;
+
+    @ResourcePath("reports/v1/report_01") <1>
+    Template report;
+
+    @Scheduled(cron="0 30 * * * ?") <2>
+    void generate() {
+        String result = report.render()
```

Comment:
Similarly to what I said above, I don't feel like this method should be called `render` since it doesn't render.

Reply:
Ok so what about `setup()`, `prepare()` or even `start()`?

Reply:
Just curious: Is a method here really necessary? Can't we just go with `report.data(..).render()`?

Reply:
I think the name you're looking for is `instantiate` ;) We can probably find a fitting name once we find an agreement for `Rendering`, I'm sure it will follow.

Reply:
> Just curious: Is a method here really necessary? Can't we just go with report.data(..).render()?

That was my first thought as well. You can start with the first parameter here and put the other ones on its result:

```java
String results = report.data("samples, service.get())
 .data("now", LocalDateTime.now())
 .render()
```

Reply:
Yes, we could and it's probably a good idea. Although in theory, you don't always need to pass data and use e.g. `@Named` beans directly in the template.

Reply:
Well we can have an overload that doesn't take data, provided we find its name ;)